Do not merge:  sql: add system.jobs table for job status and checkpoints

vivekmenezes opened the pull request:
I got back to working on this PR and have been bike-shedding on it for the last three days and feel like I can't waste more time on it. It works fine as is but I'd ask that you don't review it in detail. The main place that I need help with is the Sleep() that I've introduced to coordinate a queuing of a table split with the start of the scanners. I'd appreciate some ideas on how to add in some coordination code to eliminate that Sleep. I'd also appreciate comments on the first commit that moves the start of the scanners to the end of the server start code. Once that's done I can cleanup this PR for a full review. Thanks

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7073)

<!-- Reviewable:end -->

bdarnell commented:
How will this table be created when an existing cluster is upgraded to this version?

vivekmenezes commented:
I actually dont know when these tables are created. I'm hoping that i can
just add this table in now and we can start using it in a month or sometime
later. The other factor is that the checkpointing process can assume that
the table doesn't exist and not write checkpoints until the table is
created.

On Mon, Jun 6, 2016, 5:28 PM Ben Darnell <notifications@github.com> wrote:

> How will this table be created when an existing cluster is upgraded to
> this version?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/7073#issuecomment-224093999>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ALOpBOwXXcohXktrDS4P_SgT0dHEG59Eks5qJJEKgaJpZM4IvUfF>
> .
>

petermattis commented:
Not sure I'll have a chance to look at this tomorrow (busy meeting day), but I echo Ben's question about needing a solution for creating this table on an existing cluster. Currently, system tables are only created when a cluster is bootstrapped. And, as you've noticed, we have no mechanism for changing their schema. #5887 is an example of where a system table schema migration could be used. And now this PR is an example of where we need to create a system table if it doesn't already exist.

Some very basic brainstorming: what if the some well known node (e.g. the range leader for the first range) was responsible for creating new system tables and performing schema changes on them and we required other parts of the system to be able to operate when the new system table does not yet exist. Lots of details to flesh out with this approach. Other approaches are welcome.

vivekmenezes commented:
This PR is ready for another viewing. You can ignore the first two commits because they are in a separate PR. I've used the "IF NOT EXIST" sql clause to not require a leader election to run the table creation. I don't think we want to tackle all schema changes at this stage for we have higher priorities.

bdarnell commented:
I'm slightly worried that relying on a thundering herd of `CREATE IF NOT EXIST` statements is going to make it harder to support real migrations in the future. 

---

Review status: 0 of 24 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[server/server.go, line 499 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuXG6te2nSVXUVKnXM:-KLuXG6uKz1LmBA6c7zO:-1286249958) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/server/server.go#L499)):*
> ```Go
> 
> 	// Upgrade the system tables schema.
> 	if err := s.stopper.RunAsyncTask(func() {
> ```

Doing this asynchronously is problematic: how is code that wants to use the jobs table supposed to know whether the migration has completed or not? If we can't do the migration synchronously here, we need some other hook that can be used to wait for it to finish. 

---

*[server/server.go, line 535 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuXay2ZiwLfT8iXqN-:-KLuXay2ZiwLfT8iXqN0:1767439883) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/server/server.go#L535)):*
> ```Go
> 		r := exec.ExecuteStatements(context.Background(), session, table, nil)
> 		if err := checkQueryResults(r.ResultList, 1); err != nil {
> 			log.Warningf("err: %s, Unable to create system table: %s", err, table)
> ```

I think this should either be a fatal error or be in a retry loop. If we fail to create the table then we won't be able to do anything that uses the "jobs" system later.

---

*[server/testserver.go, line 279 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuWoBo9hPkfexDqnDJ:-KLuWoBo9hPkfexDqnDK:-605421716) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/server/testserver.go#L279)):*
> ```Go
> var allSystemTableRangeSplits = false
> 
> // AllSystemTableRangeSplits sets allSystemTableRangeSplits.
> ```

The docstring for an exported function generally shouldn't refer to unexported variables. What's going on here? Why doesn't `WaitForInitialSplits` always wait for both new and old tables to be split? 

---

*[sql/split_test.go, line 130 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuZtVoc1iOyy7O6uN0:-KLuZtVoc1iOyy7O6uN1:-887222506) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/split_test.go#L130)):*
> ```Go
> 
> 	// Verify the actual splits. Also include splits caused by the creation of new system tables.
> 	splits = []roachpb.RKey{keys.MakeTablePrefix(15), keys.MakeTablePrefix(objectID), keys.MakeTablePrefix(objectID + 1), roachpb.RKeyMax}
> ```

`15` should refer to some constant. (`keys.JobsTableID`?) Why is this table different from the other system tables in this test?

---

*[sql/sqlbase/system.go, line 76 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuRmLJ3d-35L9bUnG6:-KLuRmLJ3d-35L9bUnG7:615134466) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/system.go#L76)):*
> ```Go
> 	jobsTableSchema = `
> CREATE TABLE IF NOT EXISTS system.jobs (
> 	jobID        INT PRIMARY KEY,
> ```

How will IDs be chosen? (for both `jobID` and `typeID`) 

How will this table be accessed? I would have expected a "jobs" table to include the node id or some sort of lease for the work that is currently in progress. If we're not building a mechanism that would allow us to change this table then we need to make sure we get the schema right the first time. 

---

*[sql/sqlbase/system.go, line 123 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuSrBt7KBvOwaK77iR:-KLuSrBt7KBvOwaK77iS:-1922213154) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/system.go#L123)):*
> ```Go
> 	// 1 for the SystemDB descriptor created above, and is incremented by every
> 	// call to createSystemTable().
> 	NumSystemConfigDescriptors = 1
> ```

There are some changes here in the third commit that look like they belong in the other PR. 

---

*[sql/sqlbase/system.go, line 126 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuUdtnfpQYyu6JSyBx:-KLuUdtnfpQYyu6JSyBy:663974346) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/system.go#L126)):*
> ```Go
> 
> 	// NewSystemTablesSchema is the collection of system tables that should be
> 	// created.
> ```

Expand this comment to explain that this is for all system tables created post-beta.

---

*[sql/sqlbase/system.go, line 129 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuV1TCggXxTOqs85p0:-KLuV1TDqk7wU2vS1dZ4:-2060144405) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/system.go#L129)):*
> ```Go
> 	NewSystemTablesSchema = []string{jobsTableSchema}
> 	// NumNewSystemTablesSchema is the size of NewSystemTablesSchema.
> 	NumNewSystemTablesSchema = len(NewSystemTablesSchema)
> ```

I'm worried about the number of places where this shows up in the code. It seems that in most cases what we want is the total number of system tables, old+new, so we should have a separate variable (or function) for that. 

---

*[sql/sqlbase/system.go, line 205 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuR-uJ_jsTfJnPwFMw:-KLuR-uKz1qD5Gw-LDSX:-975660926) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/system.go#L205)):*
> ```Go
> }
> 
> func systemTableID(name string) ID {
> ```

This should probably live in the `keys` package along with the values, and it should include all system tables, not just the non-gossiped ones (or it should be limited to the "new" ones).

---

*[sql/sqlbase/table.go, line 44 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuVT6m2P8lW9Iv4zUX:-KLuVT6m2P8lW9Iv4zUY:-1245919105) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/table.go#L44)):*
> ```Go
> 	desc.Name = p.Table.Table()
> 	desc.ParentID = parentID
> 	// All system tables have their IDs pre-allocated.
> ```

Do we need to do this for new system tables? We preallocated IDs for the old ones because they were not created through the normal process (and the gossiped tables need contiguous IDs), but I think system tables created after bootstrapping (which are only used from SQL-level code) can just get their IDs in the usual fashion. 

Or maybe it's good to use preallocated ids so we can move "new" tables into the bootstrap schema when we no longer need to support upgrading from older releases. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-:-KLuXyLCkrsLRSZ9gluC:-759487710)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I agree that this mechanism is not what we should use long term. I started working on it with the idea that I get something working in a reasonable way and unfortunately it's taken more time than I anticipated. Given the priorities we have I'm afraid I'd like to not sink more time on this system table upgrade code. While I anticipate it to be important later, it's definitely not a priority now. So I'd ask that we get something working that is decent.

---

Review status: 0 of 21 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[server/server.go, line 499 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuXG6te2nSVXUVKnXM:-KLxoE_7hAuJkvGSX7Ww:1403324822) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/server/server.go#L499)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Doing this asynchronously is problematic: how is code that wants to use the jobs table supposed to know whether the migration has completed or not? If we can't do the migration synchronously here, we need some other hook that can be used to wait for it to finish. 
</details>
I suppose the code needing the table can wait for the table to appear. The schema change backfill code can start off by checking if the table exists before moving forward with a backfill.

I think it's safer to make it asynchronous. I worry about keeping this server code hanging on a dependency on the sql layer.

---

*[server/server.go, line 535 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuXay2ZiwLfT8iXqN-:-KLxojsu-G3nptSRfgDC:1820440707) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/server/server.go#L535)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I think this should either be a fatal error or be in a retry loop. If we fail to create the table then we won't be able to do anything that uses the "jobs" system later.
</details>
I've added a loop to the code to keep retrying with no backoff. I think the sql layer and below has plenty of stalls below it (raft network latency) to allow us to not have to backoff. Let me know if you think we need a retry loop here and I can add one.

---

*[server/testserver.go, line 279 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuWoBo9hPkfexDqnDJ:-KLxpSxxJyD9QwpCfAGo:-273897755) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/server/testserver.go#L279)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The docstring for an exported function generally shouldn't refer to unexported variables. What's going on here? Why doesn't `WaitForInitialSplits` always wait for both new and old tables to be split? 
</details>
The new table splits only occur if the scanner runs again. I found it rather painful to make it run again. For now WaitForInitialSplits() waits for new tables only when a fast scanner has been configured.

---

*[sql/split_test.go, line 130 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuZtVoc1iOyy7O6uN0:-KLxqFfhH48pIl8WAm3s:1916636072) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/split_test.go#L130)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> `15` should refer to some constant. (`keys.JobsTableID`?) Why is this table different from the other system tables in this test?
</details>
This is no longer needed.

---

*[sql/sqlbase/system.go, line 76 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuRmLJ3d-35L9bUnG6:-KLxqwRrnzGeg5sWB3iX:1505677776) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/system.go#L76)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> How will IDs be chosen? (for both `jobID` and `typeID`) 
> 
> How will this table be accessed? I would have expected a "jobs" table to include the node id or some sort of lease for the work that is currently in progress. If we're not building a mechanism that would allow us to change this table then we need to make sure we get the schema right the first time. 
</details>
I was thinking of creating a jobID for a schema change for instance using something like sql unique_rowid(), and was originally planning to store it in the table descriptor but I guess a job can span multiple tables so that aint gonna work general. It works for schema changes.

We have to ensure we get the schema right the first time. I suppose I should use this PR with the checkpointing PR to see how this works. Let me do that before merging this.

 

---

*[sql/sqlbase/system.go, line 126 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuUdtnfpQYyu6JSyBx:-KLxuBUJNC08UN2HVc5b:1532213654) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/system.go#L126)):*
>  this is for all system tables created post-beta.
Done

---

*[sql/sqlbase/system.go, line 129 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuV1TCggXxTOqs85p0:-KLxuDsRUr5af4vmELrG:63390062) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/system.go#L129)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I'm worried about the number of places where this shows up in the code. It seems that in most cases what we want is the total number of system tables, old+new, so we should have a separate variable (or function) for that. 
</details>
Take a look at it now.

---

*[sql/sqlbase/system.go, line 205 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuR-uJ_jsTfJnPwFMw:-KLxytRYNqSN7jnR3i50:1026971271) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/system.go#L205)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This should probably live in the `keys` package along with the values, and it should include all system tables, not just the non-gossiped ones (or it should be limited to the "new" ones).
</details>
I've kept this here because the ID type is defined in sqlbase. I've included all the system config table as well

---

*[sql/sqlbase/table.go, line 44 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuVT6m2P8lW9Iv4zUX:-KLxucKwy_uu_CwMdq74:230708424) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/table.go#L44)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Do we need to do this for new system tables? We preallocated IDs for the old ones because they were not created through the normal process (and the gossiped tables need contiguous IDs), but I think system tables created after bootstrapping (which are only used from SQL-level code) can just get their IDs in the usual fashion. 
> 
> Or maybe it's good to use preallocated ids so we can move "new" tables into the bootstrap schema when we no longer need to support upgrading from older releases. 
</details>
I think it's better to preallocate IDs because it's nice to know that they come from a certain range of numbers. We might want to treat them specially in the future like add more backups or whatever.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-:-KLxnQuj95Ssn33ZSzT7:572288898)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 8 files at r2, 8 of 20 files at r3, 12 of 14 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 20 unresolved discussions, all commit checks successful.

---

*[server/node_test.go, line 594 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy0CJwOwJ_NPlrOJa1:-KLy0CJwOwJ_NPlrOJa2:-1280936503) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/server/node_test.go#L594)):*
> ```Go
> 		t.Fatal(err)
> 	}
> ```

remove

---

*[server/server.go, line 499 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuXG6te2nSVXUVKnXM:-KLy0JZuuMiZhDsc6gnt:-792086155) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/server/server.go#L499)):*
> I suppose the code needing the table can wait for the table to appear.

Sounds like substantial baggage. I'd make this synchronous.

---

*[server/server.go, line 529 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy0gZyYEE9VmN9vdgc:-KLy0gZyYEE9VmN9vdgd:447844328) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/server/server.go#L529)):*
> ```Go
> 	for _, result := range results {
> 		if result.Err != nil {
> 			return errors.Errorf("%s", result.Err)
> ```

`return result.Err`?

---

*[server/server.go, line 541 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy0s31QsrGJxY_z_LP:-KLy0s31QsrGJxY_z_LQ:1683300842) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/server/server.go#L541)):*
> ```Go
> // creation and might not make sense to run from low level code.
> func upgradeSystemTablesSchema(exec *sql.Executor) {
> 	// System tables should be created with the highest privileges that a user
> ```

what does "highest" mean?

---

*[server/server.go, line 555 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy13yH-2RWyNDfQ-x_:-KLy13yH-2RWyNDfQ-xa:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/server/server.go#L555)):*
> ```Go
> 	}
> 
> 	// Run any schema change commands on system tables.
> ```

?

---

*[server/testserver.go, line 331 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy1Ptlzd6ajgYm6KqE:-KLy1Ptlzd6ajgYm6KqF:178357774) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/server/testserver.go#L331)):*
> ```Go
> func MakeFastScanContext() Context {
> 	c := MakeTestContext()
> 	c.ScanInterval = time.Millisecond
> ```

Instead of this, can we just force a scan on every iteration of WaitForInitialSplits (or a variant of it)? i realize this is code is not new in this change, but it's making me sad.

---

*[sql/create.go, line 247 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy1XI1DpAKnavqtYUv:-KLy1XI1DpAKnavqtYUw:-1830329395) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/sql/create.go#L247)):*
> ```Go
> 		return err
> 	}
> 	// Inherit permissions from the database descriptor.
> ```

"by default"

---

*[sql/pgwire_test.go, line 412 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy1hkpRwQOrQelTTV3:-KLy1hkpRwQOrQelTTV4:142266814) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/sql/pgwire_test.go#L412)):*
> ```Go
> 		},
> 		"SHOW TABLES FROM system": {
> 			base.Results("descriptor").Others(8),
> ```

magic number?

---

*[sql/sqlbase/privilege.go, line 229 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy2SfsKiWzcnfmV4lX:-KLy2SfsKiWzcnfmV4lY:455177946) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/sql/sqlbase/privilege.go#L229)):*
> ```Go
> 	if !ok {
> 		// User "node" has all privileges.
> 		if user == security.NodeUser {
> ```

`return user == security.NodeUser`?

---

*[sql/sqlbase/system.go, line 205 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuR-uJ_jsTfJnPwFMw:-KLy2nyhxAg7gRUY9Ove:-2118303355) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/system.go#L205)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I've kept this here because the ID type is defined in sqlbase. I've included all the system config table as well
</details>
the zero return for not-found is unfortunate. can you give it a second return?

---

*[sql/sqlbase/system.go, line 198 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy2h7rZgOfNEyDJeBo:-KLy2h7rZgOfNEyDJeBp:-395509907) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/sql/sqlbase/system.go#L198)):*
> ```Go
> 	target.AddTable(keys.LeaseTableID, leaseTableSchema)
> 	target.AddTable(keys.UITableID, uiTableSchema)
> 	// Do not add new system tables here. All new tables should be added to
> ```

briefly explain why.

---

*[storage/replicate_test.go, line 37 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy35ryTP4FNGHcPHqr:-KLy35ryTP4FNGHcPHqs:-470516354) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/storage/replicate_test.go#L37)):*
> ```Go
> 	store.SetReplicaScannerDisabled(true)
> 
> 	fastScanner := false
> ```

const

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 14 of 21 files reviewed at latest revision, 20 unresolved discussions, some commit checks pending.

---

*[server/node_test.go, line 594 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy0CJwOwJ_NPlrOJa1:-KM-aHmmXW2KTe580Wjq:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/server/node_test.go#L594)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> remove
</details>
done

---

*[server/server.go, line 499 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuXG6te2nSVXUVKnXM:-KM-aOb-aDpMTah5IasX:1901510748) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/server/server.go#L499)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> > I suppose the code needing the table can wait for the table to appear.
> 
> Sounds like substantial baggage. I'd make this synchronous.
</details>
Making it synchronous is not going to change that. I think once the server has started accepting connections jobs can start getting scheduled. I'd hate for us to get a cluster in a situation where all the nodes are stuck on Start() because of this.

---

*[server/server.go, line 529 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy0gZyYEE9VmN9vdgc:-KM-bF17ps3ee-r56vNG:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/server/server.go#L529)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `return result.Err`?
</details>
done

---

*[server/server.go, line 541 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy0s31QsrGJxY_z_LP:-KM-bhciwLAj6d5eTbxF:-720025036) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/server/server.go#L541)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what does "highest" mean?
</details>
changed to

// System tables are created by user "node" which is an internal user.

---

*[server/server.go, line 555 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy13yH-2RWyNDfQ-x_:-KM-c4e-OvS0fX7YIpjo:-1054325829) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/server/server.go#L555)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ?
</details>
removed

---

*[server/testserver.go, line 331 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy1Ptlzd6ajgYm6KqE:-KM-cHSluonEo5fuIuCo:-1681915900) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/server/testserver.go#L331)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Instead of this, can we just force a scan on every iteration of WaitForInitialSplits (or a variant of it)? i realize this is code is not new in this change, but it's making me sad.
</details>
let me look into that

---

*[sql/create.go, line 247 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy1XI1DpAKnavqtYUv:-KM-cUmjg0KwVFb9fqhi:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/sql/create.go#L247)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "by default"
</details>
done

---

*[sql/pgwire_test.go, line 412 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy1hkpRwQOrQelTTV3:-KM-cXfzmJf4FOc92QOY:-467915377) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/sql/pgwire_test.go#L412)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> magic number?
</details>
fixed

---

*[sql/sqlbase/privilege.go, line 229 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy2SfsKiWzcnfmV4lX:-KM-dyt13Kdla5W_TTIR:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/sql/sqlbase/privilege.go#L229)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `return user == security.NodeUser`?
</details>
done

---

*[sql/sqlbase/system.go, line 205 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuR-uJ_jsTfJnPwFMw:-KM-idj35Ks3cUb15gwg:-1162440976) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/system.go#L205)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> the zero return for not-found is unfortunate. can you give it a second return?
</details>
I've added a log.Warning. Unfortunately some of our unittests use system table names that are not part of the set. 

---

*[sql/sqlbase/system.go, line 198 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy2h7rZgOfNEyDJeBo:-KM-fnUE4UmOCT0k2FGx:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/sql/sqlbase/system.go#L198)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> briefly explain why.
</details>
done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 7 files at r6.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[server/server.go, line 499 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuXG6te2nSVXUVKnXM:-KM-tYmQC0HrD1NPMlg8:1868831515) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/server/server.go#L499)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Making it synchronous is not going to change that. I think once the server has started accepting connections jobs can start getting scheduled. I'd hate for us to get a cluster in a situation where all the nodes are stuck on Start() because of this.
</details>
I'd rather they were stuck on Start than be in an indeterminate state.

---

*[sql/pgwire_test.go, line 322 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KM-u-8k60Mubg_46zzY:-KM-u-8k60Mubg_46zzZ:-480817506) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3ab1499dbae7af097232e2e949e85e1d0fa90ff/sql/pgwire_test.go#L322)):*
> ```Go
> 	var base preparedQueryTest
> 
> 	// Do not count the system database descriptor and the "descriptor" descriptor; thus -2
> ```

end with a period.

---

*[sql/sqlbase/system.go, line 205 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuR-uJ_jsTfJnPwFMw:-KM-u9o50zkskl3IXyje:-1358283531) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/system.go#L205)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I've added a log.Warning. Unfortunately some of our unittests use system table names that are not part of the set. 
</details>
Can we fix the tests?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[server/server.go, line 499 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuXG6te2nSVXUVKnXM:-KM5md5IJs9cfiyWwfV5:261608874) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/server/server.go#L499)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I'd rather they were stuck on Start than be in an indeterminate state.
</details>
I'd also prefer to block in Start, if that works (I could imagine that trying to block in Start would just deadlock). But whether it's synchronous or async in Start, there should be *some* way to wait for the migrations to complete (e.g. a public channel object that is closed at the end of upgradeSystemTablesSchema)

---

*[server/server.go, line 535 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuXay2ZiwLfT8iXqN-:-KM5nQraUC_tbjf_fMTJ:-75952160) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/server/server.go#L535)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I've added a loop to the code to keep retrying with no backoff. I think the sql layer and below has plenty of stalls below it (raft network latency) to allow us to not have to backoff. Let me know if you think we need a retry loop here and I can add one.
</details>
Retry loops should always have a backoff unless it's important for them to run as quickly as possible. 

---

*[server/testserver.go, line 331 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLy1Ptlzd6ajgYm6KqE:-KM5nqbwujBaLz1fvx93:1076674825) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f846c5dd3357747d491e4e4826830de0e029221/server/testserver.go#L331)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> let me look into that
</details>
Yeah, forcing a scan sounds better to me than making the behavior of WaitForInitialSplits conditional on whether the scanner is "fast". 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073)*
<!-- Sent from Reviewable.io -->

vivekmenezes committed:
server: start scanners at the end of server initialization

This is to allow scan functionality, such as splitting ranges, to
occur only after the server has initialized.

vivekmenezes committed:
a mechanism to create system tables outside of bootstrapping process

Use the new mechanism to add the existing UITable.

bdarnell commented:
Reviewed 15 of 15 files at r8.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[server/server.go, line 629 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbjhUgRE5UMMIDB1ka:-KRbjhUgRE5UMMIDB1kb:b-ep79yd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a73af9451b6886c63d11ffc0d34592d0cb9cc1/server/server.go#L629)):*
> ```Go
> func upgradeSystemTablesSchema(exec *sql.Executor) {
> 	// System tables are created by user "node" which is an internal user.
> 	session := sql.NewSession(context.TODO(), sql.SessionArgs{User: security.NodeUser}, exec, nil)
> ```

You could easily pass a context here.

---

*[sql/create.go, line 379 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbkRxgTLy30U0OwGWS:-KRbkRxgTLy30U0OwGWT:b-owr0sk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a73af9451b6886c63d11ffc0d34592d0cb9cc1/sql/create.go#L379)):*
> ```Go
> 	}
> 
> 	if desc.ID == 0 {
> ```

Why can't we let tables created the new way have a naturally-assigned ID? 

---

*[sql/sqlbase/system.go, line 205 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuR-uJ_jsTfJnPwFMw:-KRbfuMEgbFbIk61_QKI:b-rrsm6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/system.go#L205)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Can we fix the tests?
</details>
The ID type is defined here, but the values of the ID type are defined in `keys` (as untyped constants). I'd like for this function, the ID type, and the ID constants to be defined in one package (so the constants can be properly typed). 

These tests definitely sound fishy. It seems like we should either fix the test to use defined names, or add a second return value here so that callers are forced to handle the case of undefined names. (it especially doesn't make sense to have `case "test1", "test2"` and a catch-all default)

---

*[sql/sqlbase/system.go, line 382 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbd8EuOb1L9L0mQ46i:-KRbd8EuOb1L9L0mQ46j:b-9lbm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9c3f1910c9298f7cee9c4adb0449f259f36c2af/sql/sqlbase/system.go#L382)):*
> ```Go
> 
> 	// UITable is the descriptor for the ui table.
> 	UITable = TableDescriptor{
> ```

Why is this changing? Just to test the new mechanism? I'd rather test this with a genuinely new table, because otherwise our test clusters (which will create the UI table through the new mechanism) will differ from the prod clusters (which have the preexisting UI table).

Alternately, since it's definitely nicer to not have this manual table descriptor, we should move as many of the existing tables as we can to the new mechanism. 

---

*[storage/replicate_test.go, line 37 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbi3h-WhGTXajhOtzl:-KRbi3h0cciZj6f2I1Wy:be0j7r2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a73af9451b6886c63d11ffc0d34592d0cb9cc1/storage/replicate_test.go#L37)):*
> ```Go
> 	store.SetReplicaScannerActive(false)
> 
> 	if err := server.WaitForInitialSplitsExpectedCount(store.DB(), server.ExpectedInitialRangeCount()-1); err != nil {
> ```

Why -1?

---

*[storage/store.go, line 540 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbijJMTTdE_Ausnx0m:-KRbijJMTTdE_Ausnx0n:b-ifr94t) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a73af9451b6886c63d11ffc0d34592d0cb9cc1/storage/store.go#L540)):*
> ```Go
> 	// ScannersStartNotify is a channel the scanners should wait on before
> 	// starting.
> 	ScannersStartNotify chan bool
> ```

Why `chan bool` instead of `chan struct{}`?

Why do the scanners need to wait for tables to be created? The scanners don't care about SQL-level constructs. 

---

*[storage/store.go, line 995 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbiATnTh1lbS7crCV-:-KRbiATnTh1lbS7crCV0:bmuyffa) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a73af9451b6886c63d11ffc0d34592d0cb9cc1/storage/store.go#L995)):*
> ```Go
> 			case <-s.ctx.Gossip.Connected:
> 				if s.ctx.ScannersStartNotify != nil {
> 					<-s.ctx.ScannersStartNotify
> ```

This needs to be in a `select` with `stopper.ShouldStop`.

---

*[storage/store.go, line 1008 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbiDhXoe9c7zGHg5r0:-KRbiDhXoe9c7zGHg5r1:bs4q9op) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a73af9451b6886c63d11ffc0d34592d0cb9cc1/storage/store.go#L1008)):*
> ```Go
> 			case <-s.ctx.Gossip.Connected:
> 				if s.ctx.ScannersStartNotify != nil {
> 					<-s.ctx.ScannersStartNotify
> ```

Ditto.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Thanks for taking a look. I really need advice on how I can get rid of the 100ms timeout that I've introduced in this PR before I can clean it up and send it out for a proper review. Thanks!

---

Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[sql/create.go, line 379 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbkRxgTLy30U0OwGWS:-KRdBhKpyWXLeA7ZlPvA:b-ws3rj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a73af9451b6886c63d11ffc0d34592d0cb9cc1/sql/create.go#L379)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why can't we let tables created the new way have a naturally-assigned ID? 
</details>
I'm not sure that's gonna work. We do have logic around that depends on them being within a specific range. For example we allow caching of these table descriptors without leases. 

---

*[sql/sqlbase/system.go, line 205 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuR-uJ_jsTfJnPwFMw:-KRdCKDBURAwfXksK5GZ:b-ex8x9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/system.go#L205)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The ID type is defined here, but the values of the ID type are defined in `keys` (as untyped constants). I'd like for this function, the ID type, and the ID constants to be defined in one package (so the constants can be properly typed). 
> 
> These tests definitely sound fishy. It seems like we should either fix the test to use defined names, or add a second return value here so that callers are forced to handle the case of undefined names. (it especially doesn't make sense to have `case "test1", "test2"` and a catch-all default)
</details>
I'll take out test1 and test2 in a separate PR that mrtracy can review since he added them.

---

*[sql/sqlbase/system.go, line 382 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbd8EuOb1L9L0mQ46i:-KRdCS9rpI9pzK2Cc3c_:bkafv3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9c3f1910c9298f7cee9c4adb0449f259f36c2af/sql/sqlbase/system.go#L382)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why is this changing? Just to test the new mechanism? I'd rather test this with a genuinely new table, because otherwise our test clusters (which will create the UI table through the new mechanism) will differ from the prod clusters (which have the preexisting UI table).
> 
> Alternately, since it's definitely nicer to not have this manual table descriptor, we should move as many of the existing tables as we can to the new mechanism. 
</details>
I was hoping I could move all the tables to be constructed this way but fail. A lot of the failure has to do with the fact that I don't have a controlled way to manage split (discussed in another comment). Once I have the splits under control I'll give it a try although I'm not sure if some lower level functionality depending on these tables will work. I don't want to introduce another table here; the UI table can be constructed in the new way on new clusters and old clusters can continue to use the previously created table. I don't see a conflict there.

---

*[storage/replicate_test.go, line 37 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbi3h-WhGTXajhOtzl:-KRdDSd_O0dclql2leTT:b-f9h71c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a73af9451b6886c63d11ffc0d34592d0cb9cc1/storage/replicate_test.go#L37)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why -1?
</details>
for the UITable, it needs to be cleaned up

---

*[storage/store.go, line 540 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbijJMTTdE_Ausnx0m:-KRdDZD0QghBLeJhG9C6:b3nknzi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a73af9451b6886c63d11ffc0d34592d0cb9cc1/storage/store.go#L540)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why `chan bool` instead of `chan struct{}`?
> 
> Why do the scanners need to wait for tables to be created? The scanners don't care about SQL-level constructs. 
</details>
This is the main problem that I need advice on. So by introducing the table construction through this new mechanism I've made range splitting rather unpredictable for the tests. The scanners start up in node.Start() and once they startup they run through all the splits that need to be made and then wait for a long periodic timeout before running another scan to split. To split ranges predictably for the tables created using the new mechanism, I've had to start the scanners after the sql table creation calls. But that's not enough because the queuing of replicas that need splitting is not part of the table creation transaction, so I've had to put that 100ms timeout in server.go after the tables are created and before the scanners start to get everything to work. I'd really like to get this mechanism sorted out before proceeding further on this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-:-KRdFaVNqzOrckZZ9OcC:b2no4sh)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
I'm going to hand further reviews of this PR over to the SQL team.

---

Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[sql/create.go, line 379 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbkRxgTLy30U0OwGWS:-KRhmf748ZdQLkGQNobN:bdmdgxi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a73af9451b6886c63d11ffc0d34592d0cb9cc1/sql/create.go#L379)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I'm not sure that's gonna work. We do have logic around that depends on them being within a specific range. For example we allow caching of these table descriptors without leases. 
</details>
That sounds like something we need to move away from - we'll never be able to make any changes to these tables if they don't use leases. A few tables really are special (we can't use leases for the lease table descriptor itself, for example), and the gossiped subset of system tables depend on contiguous IDs, but as much as possible we should be making our system tables (especially newly-introduced ones) act like normal tables. 

---

*[sql/sqlbase/system.go, line 382 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbd8EuOb1L9L0mQ46i:-KRhnDnMnEtoR2pCBOQQ:b3te7q1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9c3f1910c9298f7cee9c4adb0449f259f36c2af/sql/sqlbase/system.go#L382)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I was hoping I could move all the tables to be constructed this way but fail. A lot of the failure has to do with the fact that I don't have a controlled way to manage split (discussed in another comment). Once I have the splits under control I'll give it a try although I'm not sure if some lower level functionality depending on these tables will work. I don't want to introduce another table here; the UI table can be constructed in the new way on new clusters and old clusters can continue to use the previously created table. I don't see a conflict there.
</details>
There's no conflict as long as it produces *exactly* the same results, but it seems like an unnecessary risk. I'd rather make a brand-new table for testing this, or go back and convert as many of the old tables as possible to the new scheme, instead of making one table that's different because it was used for testing the transition. (unless this table happens to be the only one that can be converted to the new scheme without difficulty, then it's fine)

---

*[storage/replicate_test.go, line 37 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbi3h-WhGTXajhOtzl:-KRhntlR3H-PzyhpG6c3:bo03hh0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a73af9451b6886c63d11ffc0d34592d0cb9cc1/storage/replicate_test.go#L37)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> for the UITable, it needs to be cleaned up
</details>
Why not `- len(sqlbase.NewSystemTableSchema)`? Won't this need to be updated for each new table?

---

*[storage/store.go, line 540 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbijJMTTdE_Ausnx0m:-KRhoDtRZZMRGVMgJnS-:b-pk6ijz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a73af9451b6886c63d11ffc0d34592d0cb9cc1/storage/store.go#L540)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> This is the main problem that I need advice on. So by introducing the table construction through this new mechanism I've made range splitting rather unpredictable for the tests. The scanners start up in node.Start() and once they startup they run through all the splits that need to be made and then wait for a long periodic timeout before running another scan to split. To split ranges predictably for the tables created using the new mechanism, I've had to start the scanners after the sql table creation calls. But that's not enough because the queuing of replicas that need splitting is not part of the table creation transaction, so I've had to put that 100ms timeout in server.go after the tables are created and before the scanners start to get everything to work. I'd really like to get this mechanism sorted out before proceeding further on this PR.
</details>
Triggering splits at table creation is tricky because the table creation doesn't actually touch the range where the data will live. The only way I can see is to issue AdminSplit calls from sql when CREATE TABLE is executed. (it doesn't need to be tied to the transaction; you can split as soon as the id is allocated)

---

*[sql/sqlbase/table.go, line 44 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuVT6m2P8lW9Iv4zUX:-KRhqRjx_uVKgMznaFLX:bkl6269) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/table.go#L44)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I think it's better to preallocate IDs because it's nice to know that they come from a certain range of numbers. We might want to treat them specially in the future like add more backups or whatever.
</details>
Why would we want to do that by ID instead of by namespace?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-:-KRhqbYXdcjhhsAwbmbl:b-onxi1x)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Thanks for taking a look. I've figured out a way to move forward with this PR so I'm happy. 

---

Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[sql/create.go, line 379 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbkRxgTLy30U0OwGWS:-KRi_sjvXK458IjH4JpD:b-1mcf7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a73af9451b6886c63d11ffc0d34592d0cb9cc1/sql/create.go#L379)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> That sounds like something we need to move away from - we'll never be able to make any changes to these tables if they don't use leases. A few tables really are special (we can't use leases for the lease table descriptor itself, for example), and the gossiped subset of system tables depend on contiguous IDs, but as much as possible we should be making our system tables (especially newly-introduced ones) act like normal tables. 
</details>
true. I'll write up an issue on this for the future, but don't intend to add that to this PR.

---

*[storage/store.go, line 540 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KRbijJMTTdE_Ausnx0m:-KRiZnmwZxz04MNeCe1q:b-9wr8kh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0a73af9451b6886c63d11ffc0d34592d0cb9cc1/storage/store.go#L540)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Triggering splits at table creation is tricky because the table creation doesn't actually touch the range where the data will live. The only way I can see is to issue AdminSplit calls from sql when CREATE TABLE is executed. (it doesn't need to be tied to the transaction; you can split as soon as the id is allocated)
</details>
I figured out how to do this finally! It was as simple as changing TestServer to no upgrade the tables in Server.Start(), and then wait for all the splits and run the table upgrade process separately. A few tests that depend on the splits being accurate can run more frequent scans. I'm gonna move forward with this approach and see what it produces.

---

*[sql/sqlbase/table.go, line 44 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-KLuVT6m2P8lW9Iv4zUX:-KRi_ZyEDTIxQHUa-mMy:bdaek36) ([raw file](https://github.com/cockroachdb/cockroach/blob/e546ebd12dd9124f5fdf76e8b08436ccc93524b8/sql/sqlbase/table.go#L44)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why would we want to do that by ID instead of by namespace?
</details>
good point

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7073#-:-KRia8upgWJImhxoNg50:b8wcy1o)*
<!-- Sent from Reviewable.io -->
